includes/TaxCalculation/class-cart-tax-applicator.php

```diff
@@ -0,0 +1,331 @@
+			$total_taxes           = $this->apply_line_total_tax( $item_key, $item, $wc_rate );
+			$merged_subtotal_taxes = $this->merge_tax_arrays( $merged_subtotal_taxes, $subtotal_taxes );
+			$merged_total_taxes    = $this->merge_tax_arrays( $merged_total_taxes, $total_taxes );
+		}
+
+		$this->items_taxes = $merged_total_taxes;
+		$this->cart->set_subtotal_tax( wc_remove_number_precision( array_sum( $merged_subtotal_taxes ) ) );
+		$this->cart->set_cart_contents_tax( array_sum( wc_remove_number_precision_deep( $merged_total_taxes ) ) );
+		$this->cart->set_cart_contents_taxes( wc_remove_number_precision_deep( $merged_total_taxes ) );
+	}
+
+	/**
+	 * Gets the WooCommerce structured rates for a line item.
+	 *
+	 * @param string $item_key key of item in the cart.
+	 * @param array  $item Line item.
+	 *
+	 * @return array[]
+	 * @throws Exception When no line item is found on the tax details for the given key.
+	 */
+	private function get_woocommerce_rate( $item_key, $item ): array {
+		$tax_detail_line_item = $item['data']->get_id() . '-' . $item_key;
+		$rate                 = $this->get_rate_with_key( $tax_detail_line_item );
+		return $this->build_woocoommerce_rate( $rate );
+	}
+
+	/**
+	 * Gets the TaxJar rate from tax details for given key.
+	 *
+	 * @param string $tax_details_line_key line item id from TaxJar API request.
+	 *
+	 * @return float|int
+	 * @throws Exception When no line item is found on the tax details for the given key.
+	 */
+	private function get_rate_with_key( $tax_details_line_key ) {
+		$tax_detail_line_item = $this->tax_details->get_line_item( $tax_details_line_key );
+
+		if ( false === $tax_detail_line_item ) {
+			throw new Exception( 'Line item not present in tax details.' );
+		}
+
+		return 100 * $tax_detail_line_item->get_tax_rate();
+	}
+
+	/**
+	 * Builds a WooCommerce structured tax rate.
+	 *
+	 * @param float  $rate Tax rate.
+	 * @param string $shipping_taxable yes if shipping is taxable, no if not.
+	 *
+	 * @return array[]
+	 */
+	private function build_woocoommerce_rate( $rate, $shipping_taxable = 'no' ): array {
+		return array(
+			0 => array(
+				'rate'     => $rate,
+				'label'    => '',
+				'shipping' => $shipping_taxable,
+				'compound' => 'no',
+			),
+		);
+	}
+
+	/**
+	 * Applies WooCommerce structured tax rate to line subtotals.
+	 *
+	 * @param string $item_key key of item in cart.
+	 * @param array  $item item in cart.
+	 * @param array  $wc_rate WooCommerce structure rate.
+	 *
+	 * @return array
+	 */
+	private function apply_line_subtotal_tax( $item_key, $item, $wc_rate ): array {
+		$item_subtotal  = wc_add_number_precision( $item['data']->get_price() * $item['quantity'], false );
+		$subtotal_taxes = WC_Tax::calc_tax( $item_subtotal, $wc_rate );
+		$subtotal_tax   = array_sum( array_map( array( $this, 'round_line_tax' ), $subtotal_taxes ) );
+		$this->cart->cart_contents[ $item_key ]['line_tax_data']     = array( 'subtotal' => wc_remove_number_precision_deep( $subtotal_taxes ) );
```

Comment:
Do you think there's any benefit in having the `line_tax_data` array instantiated if needed in this method & `apply_line_total_tax`, so the order of the function calls don't matter?

Reply:
I see how it is procedurally safe as-is, but I think @poops has a point.

If it was something like:

`$this->cart->cart_contents[ $item_key ]['line_tax_data']['subtotal'] = wc_remove_number_precision_deep( $subtotal_taxes );
 `

as you do on line 175 below, then if these methods were called in a different order, it wouldn't be destructive to the 'total' value set in `apply_line_total_tax`

Reply:
I agree this is a better way and have fixed it. Thanks!